ML rules fixes, new rule for msgpack-numpy

dhalf opened the pull request:
This PR contains:
- Some changes/fixes to ML-related Python rules and/or their associated test file: lxml-in-pandas, numpy-in-pytorch-modules and pickles-in-pytorch
- A new rule to spot the usage of msgpack-numpy, a Python library that prepares numpy objects to be serialized with msgpack but that is backed by pickle. The rule excludes normal uses of msgpack (without the numpy translation layer).

Commit:
improves some rules and tests

Commit:
msgpack-numpy rule

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/trailofbits/semgrep-rules?pullRequest=39) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/trailofbits/semgrep-rules?pullRequest=39) before we can accept your contribution.<br/><hr/>**Lucas Bourtoule** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/trailofbits/semgrep-rules?pullRequest=39) it.</sub>

Commit:
new rule for pandas eval functions

GrosQuildu commented:
Thanks! Will review soon. Seems like you have to sign the CLA?

dhalf commented:
> Thanks! Will review soon. Seems like you have to sign the CLA?

Hi! I did sign twice but this doesn't seem to be reflected here... Any idea?

GrosQuildu committed:
Merge branch 'main' into lucas

dhalf committed:
keras and tf loading functions

dhalf committed:
signed commit

dhalf committed:
Merge branch 'lucas2' into lucas

dhalf committed:
new rules: keras and tf loading

dhalf commented:
I added two new rules to detect the use of Keras and Tensorflow loading functions. We reject all calls except those with a string literal (or constant) as file path. I also added a warning rule to deprecate the usage of pickle and hdf5 for Keras. The preferred format is now .keras which comes with checks that mitigate the risk of arbitrary code execution.

GrosQuildu commented:
Overall looks good. Need @suhacker1 to review these (or somebody from ML). Then I can review from the semgrep/cq perspective.

suhacker1 reviewed (dismissed):
From the ML-perspective, everything looks good to me!

GrosQuildu committed:
fix formatting, less FPs for pandas-eval

GrosQuildu committed:
pandas-eval metadata

GrosQuildu committed:
merge conflicts

mschwager commented on python/pandas-eval.yaml:
Is `f""` intended? We're trying to remove only empty f-strings?

GrosQuildu replied:
It is intended. We want to remove all static strings, including empty strings. Semgrep seems to have a bug - according to semgrep:
* `f" expr "` - is static string
* `f" {expr} "` - is dynamic string
* `f""` - is dynamic string

Will add a testcase for that and make issue in the upstream r2c

GrosQuildu replied:
https://github.com/semgrep/semgrep/issues/10047

GrosQuildu replied:
Issue fixed in upstream, we can fix that now.

mschwager commented on python/numpy-in-pytorch-modules.yaml:
```diff
@@ -15,7 +15,9 @@ rules:
-      - pattern: $RESULT = numpy.$FUNCTION(...)
+      - pattern-either:
+          - pattern: numpy.$FN(...)
+          - pattern: numpy.$MOD.$FN(...)
```
You can also use something like `numpy. ... .$FN(...)` if you'd like to match arbitrary depth.

GrosQuildu replied:
Good point, fixed

mschwager commented on python/pickles-in-keras-deprecation.yaml:
I think this is great as-is. I tend to write rules of this form like this, which may be slightly more ergonomic:

```yaml
                - pattern: tensorflow.keras.saving.load_model("$FILE")
            - metavariable-regex:
                metavariable: $FILE
                regex: ".*\.keras"
```

GrosQuildu replied:
fixed

mschwager reviewed (commented):
LGTM! I left a few minor comments 👍

GrosQuildu committed:
pandas-eval, add testcase for empty f-string

GrosQuildu committed:
numpy-in-pytorch-modules, generic match

GrosQuildu committed:
pickles-in-keras-deprecation - regex
